src/ghga_service_commons/utils/crypt4gh.py

```diff
@@ -0,0 +1,193 @@
+    """
+    keys = [(0, private_key, public_key)]
+    header_content = crypt4gh.header.make_packet_data_enc(0, file_secret)
+    header_packets = crypt4gh.header.encrypt(header_content, keys)
+    return crypt4gh.header.serialize(header_packets)
+
+
+@key_secret_decoder
+def decrypt_file(
+    *,
+    input_path: Union[str, Path],
+    output_path: Union[str, Path],
+    private_key: Union[str, bytes],
+) -> None:
+    """Decrypt a Crypt4GH encrypted file.
+
+    Private key should be passed as base64 encoded string or raw bytes.
+    """
+    if isinstance(input_path, str):
+        input_path = Path(input_path)
+
+    if isinstance(output_path, str):
+        output_path = Path(output_path)
+
+    keys = [(0, private_key, None)]
+    with input_path.open("rb") as infile, output_path.open("wb") as outfile:
+        crypt4gh.lib.decrypt(keys=keys, infile=infile, outfile=outfile)
+
+
+@key_secret_decoder
+def extract_file_secret(
+    *,
+    encrypted_header: Union[str, bytes],
+    private_key: Union[str, bytes],
+    public_key: Union[str, bytes],
+) -> bytes:
+    """Extract symmetric secret from Crypt4GH header/envelope.
+
+    Arguments should be passed as base64 encoded string or raw bytes.
+
+    Returns:
+        bytes: symmetric files secret from header.
+            Only one secret is supported for now.
+    """
+    # (method - only 0 supported for now, private_key, public_key)
+    keys = [(0, private_key, None)]
+    if isinstance(encrypted_header, str):
+        encrypted_header = base64.b64decode(encrypted_header)
+
+    infile = io.BytesIO(encrypted_header)
+    session_keys, _ = crypt4gh.header.deconstruct(
+        infile=infile, keys=keys, sender_pubkey=public_key
+    )
+
+    return session_keys[0]
+
+
+def generate_keypair() -> Crypt4GHKeyPair:
+    """Generate a new Crypt4GH keypair."""
+    sk_file, sk_path = mkstemp(prefix="private", suffix=".key")
+    pk_file, pk_path = mkstemp(prefix="public", suffix=".key")
+
+    # Crypt4GH does not reset the umask it sets, so we need to deal with it
+    original_umask = os.umask(0o022)
+    passphrase = os.urandom(32).hex().encode()
+    c4gh.generate(seckey=sk_file, pubkey=pk_file, passphrase=passphrase)
+    public_key = get_public_key(pk_path)
+    private_key = get_private_key(sk_path, lambda: passphrase.decode())
+    os.umask(original_umask)
+
+    Path(pk_path).unlink()
+    Path(sk_path).unlink()
+
+    return Crypt4GHKeyPair(private=private_key, public=public_key)
+
+
+@key_secret_decoder
+def random_encrypted_content(
+    file_size: int, private_key: Union[str, bytes], public_key: Union[str, bytes]
+) -> RandomEncryptedData:
+    """Create an in-memory file with random content that is Crypt4GH encrypted."""
+    encrypted_file = io.BytesIO()
+
+    with big_temp_file(file_size) as raw_file:
+        # rewind input file for reading
+        raw_file.seek(0)
+        true_size = len(raw_file.read())
```

Comment:
```suggestion
        true_size = raw_file.tell()
        # rewind input file for reading
        raw_file.seek(0)
```

should be a bit faster

Reply:
Thanks for catching that one, just copied that from a service and did not take a second look